image/descriptor.go

```diff
@@ -73,7 +73,17 @@ func findDescriptor(w walker, name string) (*descriptor, error) {
 	}
 }
 
-func (d *descriptor) validate(w walker) error {
+func (d *descriptor) validate(w walker, mts []string) error {
```

Comment:
I don't think this is the right API.  There interesting descriptor-validation cases for each descriptor property:

a. The referenced blob is exists in (or is missing from) CAS (`digest`).
b. The media type can (or cannot) be validated by this package (`mediaType`).
c. The referenced blob exists in CAS with the right (or wrong) size (`size`).

Those also apply recursively, if you want to walk the Merkle tree of blob ancestors.

If the user wants to validate a descriptor, they should just be able to call `image.ValidateDescriptor(walker, descriptor, recursive)` or [some such][1].  If they *also* want to restrict the media type of the root descriptor, they can do that outside of `ValidateDescriptor` (because there's not a convenient way to specify those restrictions recursively, i.e. “if an ancestor contains a manifest, only allow layer types from {set}”).

If you have an image-layout with a ref with an image/png media type pointing to a valid PNG blob, I `oci-image-tool validate image.tar png-ref` could do any of:

i. Print something like “unrecognized media type 'image/png'" and exit nonzero (“I don't know”).
ii. Confirm that the referenced blob is a PNG of the right size and exit zero (“This is valid”).
iii. Print something like “invalid media type 'image/png'" and exit nonzero (“This is not valid”).

It *is* a valid reference (so (i) and (ii) are ok, but (iii) is not).  The PNG is just not a reference that `oci-image-tool unpack` can handle.  I think the right API for that is to have a separate “unpackable” check bound to a flag.  So the following all pass:

    $ oci-image-tool validate --recursive --unpackable image.tar $VALID_MANIFEST_REF
    $ oci-image-tool validate --unpackable image.tar $BUSTED_ANCESTOR_MANIFEST_REF
    $ oci-image-tool validate --recursive image.tar $VALID_LAYER_REF
    $ oci-image-tool validate --recursive image.tar $VALID_CONFIG_REF
    $ oci-image-tool validate image.tar $VALID_MANIFEST_REF
    $ oci-image-tool validate image.tar $BUSTED_ANCESTOR_MANIFEST_REF
    $ oci-image-tool validate image.tar $VALID_LAYER_REF
    $ oci-image-tool validate image.tar $VALID_CONFIG_REF

and the following all fail:

    $ oci-image-tool validate --recursive --unpackable image.tar $BUSTED_ANCESTOR_MANIFEST_REF
    $ oci-image-tool validate --recursive --unpackable image.tar $VALID_LAYER_REF
    $ oci-image-tool validate --recursive --unpackable image.tar $VALID_CONFIG_REF
    $ oci-image-tool validate --unpackable image.tar $VALID_LAYER_REF
    $ oci-image-tool validate --unpackable image.tar $VALID_CONFIG_REF

[1]: https://github.com/opencontainers/image-spec/pull/262#issuecomment-245187243